ledger/catchpointtracker_test.go

```diff
@@ -413,3 +416,202 @@ func TestCatchpointTrackerPrepareCommit(t *testing.T) {
 		}
 	}
 }
+
+// blockingTracker is a testing tracker used to test "what if" a tracker would get blocked.
+type blockingTracker struct {
+	postCommitUnlockedEntryLock   chan struct{}
+	postCommitUnlockedReleaseLock chan struct{}
+	postCommitEntryLock           chan struct{}
+	postCommitReleaseLock         chan struct{}
+	committedUpToRound            int64
+}
+
+// loadFromDisk is not implemented in the blockingTracker.
+func (bt *blockingTracker) loadFromDisk(ledgerForTracker, basics.Round) error {
+	return nil
+}
+
+// newBlock is not implemented in the blockingTracker.
+func (bt *blockingTracker) newBlock(blk bookkeeping.Block, delta ledgercore.StateDelta) {
+}
+
+// committedUpTo in the blockingTracker just stores the committed round.
+func (bt *blockingTracker) committedUpTo(committedRnd basics.Round) (minRound, lookback basics.Round) {
+	atomic.StoreInt64(&bt.committedUpToRound, int64(committedRnd))
+	return committedRnd, basics.Round(0)
+}
+
+// produceCommittingTask is not used by the blockingTracker
+func (bt *blockingTracker) produceCommittingTask(committedRound basics.Round, dbRound basics.Round, dcr *deferredCommitRange) *deferredCommitRange {
+	return dcr
+}
+
+// prepareCommit, is not used by the blockingTracker
+func (bt *blockingTracker) prepareCommit(*deferredCommitContext) error {
+	return nil
+}
+
+// commitRound is not used by the blockingTracker
+func (bt *blockingTracker) commitRound(context.Context, *sql.Tx, *deferredCommitContext) error {
+	return nil
+}
+
+// postCommit implements entry/exit blockers, designed for testing.
+func (bt *blockingTracker) postCommit(ctx context.Context, dcc *deferredCommitContext) {
+	if dcc.isCatchpointRound && dcc.catchpointLabel != "" {
+		bt.postCommitEntryLock <- struct{}{}
+		<-bt.postCommitReleaseLock
+	}
+}
+
+// postCommitUnlocked implements entry/exit blockers, designed for testing.
+func (bt *blockingTracker) postCommitUnlocked(ctx context.Context, dcc *deferredCommitContext) {
+	if dcc.isCatchpointRound && dcc.catchpointLabel != "" {
+		bt.postCommitUnlockedEntryLock <- struct{}{}
+		<-bt.postCommitUnlockedReleaseLock
+	}
+}
+
+// handleUnorderedCommit is not used by the blockingTracker
+func (bt *blockingTracker) handleUnorderedCommit(uint64, basics.Round, basics.Round) {
+}
+
+// close is not used by the blockingTracker
+func (bt *blockingTracker) close() {
+}
+
+func TestCatchpointTrackerNonblockingCatchpointWriting(t *testing.T) {
```

Comment:
I tried to get this test to fail by reverting the fixes as below:
```diff
~/ga/ledger$ go test -v -run TestCatchpointTrackerNonblockingCatc
=== RUN   TestCatchpointTrackerNonblockingCatchpointWriting
--- PASS: TestCatchpointTrackerNonblockingCatchpointWriting (5.11s)
PASS
ok  	github.com/algorand/go-algorand/ledger	5.396s
~/ga/ledger$ git diff
diff --git a/ledger/catchpointtracker.go b/ledger/catchpointtracker.go
index ab03db4ef..3ca561a42 100644
--- a/ledger/catchpointtracker.go
+++ b/ledger/catchpointtracker.go
@@ -393,9 +393,7 @@ func (ct *catchpointTracker) postCommit(ctx context.Context, dcc *deferredCommit
    ct.roundDigest = ct.roundDigest[dcc.offset:]

    ct.catchpointsMu.Unlock()
-}

-func (ct *catchpointTracker) postCommitUnlocked(ctx context.Context, dcc *deferredCommitContext) {
    if dcc.isCatchpointRound && ct.archivalLedger && dcc.catchpointLabel != "" {
        // generate the catchpoint file. This need to be done inline so that it will block any new accounts that from being written.
        // the generateCatchpoint expects that the accounts data would not be modified in the background during it's execution.
@@ -410,6 +408,9 @@ func (ct *catchpointTracker) postCommitUnlocked(ctx context.Context, dcc *deferr
    }
 }

+func (ct *catchpointTracker) postCommitUnlocked(ctx context.Context, dcc *deferredCommitContext) {
+}
+
 // handleUnorderedCommit is a special method for handling deferred commits that are out of order.
 // Tracker might update own state in this case. For example, account updates tracker cancels
 // scheduled catchpoint writing that deferred commit.
diff --git a/ledger/tracker.go b/ledger/tracker.go
index 44a255b03..4a3ff2e13 100644
--- a/ledger/tracker.go
+++ b/ledger/tracker.go
@@ -328,8 +328,6 @@ func (tr *trackerRegistry) scheduleCommit(blockqRound, maxLookback basics.Round)
    }
    if cdr != nil {
        dcc.deferredCommitRange = *cdr
-   } else {
-       dcc = nil
    }

    tr.mu.RLock()
```
but it still passed — 

Reply:
so, rolling the code changes back is not enough, since the issue was that it was taking a long time. To emulate that, I've created the blockingTracker. If you'll apply the same logic you've tried above to the blockingTracker, you should be able to reproduce the fail case.

Reply:
ohhhhhhh I see you are not hooking up the test to the catchpointTracker directly. OK!